### PR TITLE
Support CDMS all species option; fix format for CDMS linelist reading; fix CDMS quantum numbers parsing

### PR DESCRIPTION
**Pull request by ParfenovS**
The default value for molecule parameter of query_lines function is 'All'. One, therefore, can expect that the following example code should work and return the lines of all species:
```
import astropy.units as u
from astroquery.linelists.cdms import CDMS
CDMS.query_lines(min_frequency=100.3 * u.GHz,max_frequency=100.5 * u.GHz,min_strength=-5)
```
However, this code does not work. This PR adds support for such a use case and makes the code above work. The returned line list can include the species with bad formatting. This PR adds a filter to exclude such species from the list (there are species with bad formatting in the frequency range in the example code above). Maybe it can be a good idea to warn a user that some lines can be excluded from the query results (I think one can discuss on how to do it in a proper way). This PR also adds some test of the new functionality. This test does not follow a pattern of other tests in that it does not utilize the the row number to check the test query result. The intent is to avoid the test breakage due to CDMS data updates.

[PR2385](https://github.com/astropy/astroquery/pull/2385) fixed some errors of line list parsing. However, some errors still remained. The starts value for MOLWT field should be decreased as this field occupies a larger number of characters that it is assumed in the current code version. An example is contained in the results of the example code above, where MOLWT field has an extra character for 1-CN-2-CCH-C6H4.
In the current code version, parse_letternumber function does not work correctly for parameters like a0, b0 etc. as it should according to CDMS documentation.
This PR fixes both these parsing errors.

### CODE REVIEW

**Comment by keflavich** [comment]
Thanks for these fixes.

Regarding the failed example cases, I'll want to more carefully review and note them in tests.  I'll do that now, just letting you know it's work in progress.

**Comment by keflavich** [comment]
I'm trying to hunt down a molecule that actually uses the negative values for quantum numbers. Any ideas for a good one to use?

**Comment by keflavich** [comment]
I built more tests, and in doing so, discovered more problems & corner cases.  I have the feeling that CDMS is changing their fixed formatting on me every week, but the lack of test coverage means I have no evidence for that.

**Comment by ParfenovS** [comment]
CH3CN is a good example of molecule, where lower case characters are used to signal negative quantum numbers smaller than –9.

**Review by bsipocz — COMMENTED** [review]
Needs a changelog, also I see test failures with ``pytest -P linelists -R``. Could you @keflavich have another look?

(also, I see a couple of commits even with a "needs to be squashed" message, so do go ahead and give it a squash, too :) )

**Comment by keflavich** [comment]
I swear CDMS is trolling me.... this was working 3 weeks ago.  I'll look more closely again.

**Comment by bsipocz** [comment]
> I swear CDMS is trolling me.... this was working 3 weeks ago. I'll look more closely again.

Yeap, I'm sorry that it took some for me to get back to this for a final review. On the bright side, it's not `main` that is broken now, only the PR 😅

File: astroquery/linelists/cdms/core.py [hunk 1]
```diff
@@ -278,40 +278,48 @@ def _parse_result(self, response, *, verbose=False):
...
+            # happen, but, well, I anticipate it will.
+            if self._last_query_temperature == 0:
+                result.rename_column('LGINT', 'LGAIJ')
+                result['LGAIJ'].unit = u.s**-1
+            else:
+                result['LGINT'].unit = u.nm**2 * u.MHz
+            result['ELO'].unit = u.cm**(-1)
+        except ValueError as ex:
+            # Give users a more helpful exception when parsing fails
+            original_message = str(ex)
```
Comment (bsipocz) [inline astroquery/linelists/cdms/core.py]: Why do you keep this variable if not using it?

File: astroquery/linelists/cdms/core.py [hunk 2]
```diff
@@ -421,25 +429,24 @@ def get_molecule(self, molecule_id, *, cache=True, return_response=False):
                                  timeout=self.TIMEOUT, cache=cache)
         if return_response:
             return response
-        result = self._parse_cat(response)
+        result = self._parse_cat(response.text)
```
Comment (bsipocz) [inline astroquery/linelists/cdms/core.py:450]: shall we raise_for_status prior to this? Or maybe can we even upstream that raise for status and do it inside `_request()`? -- that case should/would be a follow-up rather than in this PR.

File: astroquery/linelists/cdms/core.py [hunk 3]
```diff
@@ -421,25 +429,24 @@ def get_molecule(self, molecule_id, *, cache=True, return_response=False):
...
 
         species_table = self.get_species_table()
         result.meta = dict(species_table.loc[int(molecule_id)])
 
         return result
 
-    def _parse_cat(self, response, *, verbose=False):
+    def _parse_cat(self, text, *, verbose=False):
```
Comment (bsipocz) [inline astroquery/linelists/cdms/core.py:457]: nitpick, but maybe call it text_response :)

File: astroquery/linelists/cdms/core.py [hunk 4]
```diff
@@ -387,7 +418,7 @@ def tryfloat(x):
 
         return result
 
-    def get_molecule(self, molecule_id, *, cache=True):
+    def get_molecule(self, molecule_id, *, cache=True, return_response=False):
```
Comment (bsipocz) [inline astroquery/linelists/cdms/core.py:420]: while you're at it adding a new kwarg, add docstring for the method.

File: astroquery/linelists/cdms/core.py [hunk 5]
```diff
@@ -486,13 +521,16 @@ def parse_letternumber(st):
     From the CDMS docs:
     "Exactly two characters are available for each quantum number. Therefore, half
     integer quanta are rounded up ! In addition, capital letters are used to
-    indicate quantum numbers larger than 99. E. g. A0 is 100, Z9 is 359. Small
-    types are used to signal corresponding negative quantum numbers."
+    indicate quantum numbers larger than 99. E. g. A0 is 100, Z9 is 359. Lower case characters
+    are used similarly to signal negative quantum numbers smaller than –9. e. g., a0 is –10, b0 is –20, etc."
     """
+    if np.ma.is_masked(st):
+        return -999999
```
Comment (bsipocz) [inline astroquery/linelists/cdms/core.py:543]: document this in the docstring

**Review by bsipocz — COMMENTED** [review]
There are some doctest failures, but overall this is looking good (well, I trust your parser and tests, I didn't dive into that part of the code during review). 

Please fix the doctests and this can be merged.

File: astroquery/linelists/cdms/core.py (hunk 1 above)
Comment (keflavich) [reply astroquery/linelists/cdms/core.py]: leftovers that I forgot to delete; good catch

File: astroquery/linelists/cdms/core.py (hunk 2 above)
Comment (keflavich) [reply astroquery/linelists/cdms/core.py:450]: yes, just adding the raise_for_status here for now.  Doing it in request is an idea, but then it changes the behavior of `request` from being near-equivalent to `request.get`/`request.post` to being very different - I don't like that, it will have a lot of secondary effects that I expect would make debugging harder.

File: astroquery/linelists/cdms/core.py (hunk 3 above)
Comment (keflavich) [reply astroquery/linelists/cdms/core.py:457]: this function works on any `.cat` file from CDMS, so I'd like to consider making it public and leaving this as-is.  Maybe that can be a different PR though.

File: astroquery/linelists/cdms/core.py (hunk 4 above)
Comment (keflavich) [reply astroquery/linelists/cdms/core.py:420]: :+1:
